src/QsCompiler/SyntaxProcessor/CapabilityInference/ResultAnalyzer.fs

```diff
@@ -0,0 +1,186 @@
+    | ReturnInDependentBranch
+
+    /// A set statement in the branch of an if statement whose condition depends on a result, but which is reassigning a
+    /// variable declared outside the block.
+    | SetInDependentBranch of name: string
+
+type Context =
+    {
+        CallableKind: QsCallableKind
+        FrozenVars: string Set
+        InCondition: bool
+    }
+
+let createPattern kind range =
+    let opacity =
+        match kind with
+        | ConditionalEqualityInOperation -> ResultOpacity.controlled
+        | UnrestrictedEquality
+        | ReturnInDependentBranch
+        | SetInDependentBranch _ -> ResultOpacity.transparent
+
+    let capability = TargetCapability.withResultOpacity opacity TargetCapability.bottom
+
+    let diagnose (target: Target) =
+        let range = QsNullable.defaultValue Range.Zero range
+
+        match kind with
+        | _ when TargetCapability.subsumes target.Capability capability -> None
+        | ConditionalEqualityInOperation
+        | UnrestrictedEquality ->
+            let code =
+                if target.Capability.ResultOpacity = ResultOpacity.opaque then
+                    ErrorCode.UnsupportedResultComparison
+                else
+                    ErrorCode.ResultComparisonNotInOperationIf
+
+            Some(code, [ target.Name ])
+        | ReturnInDependentBranch ->
+            if target.Capability.ResultOpacity = ResultOpacity.opaque then
+                None
+            else
+                Some(ErrorCode.ReturnInResultConditionedBlock, [ target.Name ])
+        | SetInDependentBranch name ->
+            if target.Capability.ResultOpacity = ResultOpacity.opaque then
+                None
+            else
+                Some(ErrorCode.SetInResultConditionedBlock, [ name; target.Name ])
+        |> Option.map (fun (code, args) -> QsCompilerDiagnostic.Error(code, args) range)
+
+    {
+        Capability = capability
+        Diagnose = diagnose
+        Properties = ()
+    }
+
+/// Returns true if the expression is an equality or inequality comparison between two expressions of type Result.
+let isResultEquality expression =
+    let validType =
+        function
+        | InvalidType -> None
+        | kind -> Some kind
+
+    let binaryType lhs rhs =
+        validType lhs.ResolvedType.Resolution |> Option.defaultValue rhs.ResolvedType.Resolution
+
+    // This assumes that:
+    // - Result has no subtype that supports equality comparisons.
+    // - Compound types containing Result (e.g., tuples or arrays of results) do not support equality comparison.
+    match expression.Expression with
+    | EQ (lhs, rhs)
+    | NEQ (lhs, rhs) -> binaryType lhs rhs = Result
+    | _ -> false
+
+let freezeContext dependsOnResult (block: QsPositionedBlock) (context: _ ref) =
+    let knownVars = Seq.map (fun v -> v.VariableName) block.Body.KnownSymbols.Variables |> Set.ofSeq
+    let frozenVars = if dependsOnResult then knownVars else context.Value.FrozenVars
```

Comment:
What are frozen variables?

Reply:
I added a comment to the field in `Context` describing it. It's the term I use for variables declared outside of an result conditional branch while you are inside that branch. For example:

```qsharp
mutable x = 0;
if M(q) == One {
    // x is frozen
}
```

Reply:
Got it, thanks.